internal/trie/node/header.go

```diff
@@ -111,12 +130,12 @@ func decodeHeader(reader io.Reader) (variant byte,
 			// maximum uint16 value; therefore if we overflowed, we went over
 			// this maximum.
 			overflowed := maxPartialKeyLength - previousKeyLength + partialKeyLength
-			return 0, 0, fmt.Errorf("%w: overflowed by %d", ErrPartialKeyTooBig, overflowed)
+			return variant{}, 0, fmt.Errorf("%w: overflowed by %d", ErrPartialKeyTooBig, overflowed)
 		}
 
 		if buffer[0] < 255 {
 			// the end of the partial key length has been reached.
-			return variant, partialKeyLength, nil
+			return nodeVariant, partialKeyLength, nil
 		}
```

Comment:
I think there is a small problem with this guard loop. We need to move this guard `if buffer[0] < 255` before adding the value to the `partialKeyLength` otherwise it will include the wrong value in the partialKeyLength and then exits.

```go
_, err = reader.Read(buffer)
...

if buffer[0] < 255 {
   // the end of the partial key length has been reached.
   return nodeVariant, partialKeyLength, nil
}

// you can add 255 directly as they do in substrate 
// https://github.com/paritytech/substrate/blob/e354ae949964001a39a4d91761d0973775f5d7c4/primitives/trie/src/node_header.rs#L171
partialKeyLength += 255
```

Reply:
Good catch! But I think given the way we solved it this case is covered.

What we want to check is that we don't have more bytes to add (when the last read byte is < 255), for example if we have `0b0011_1111 + 0b1111_1111 + 0b1111_0000` we are adding it without issues since we are adding and then checking what is the last byte we added. Actually that's a current unit test and is passing without issues.

Do you have an example where this could fail?

Reply:
Oh true!! nevermind, I read the spec and they it says:
> Any value smaller than the maximum possible value ... implies that the next value should not be added to the length

And I notice that in the substrate code if `n < 255` then they include the n in the returned value `result + n + 1`
https://github.com/paritytech/substrate/blob/e354ae949964001a39a4d91761d0973775f5d7c4/primitives/trie/src/node_header.rs#L171, mybad 😢

Reply:
Exactly, it's almost the same but in a different way 😄 
I'll mark this conversation as resolved! thanks Eclesio!